src/Foundatio/Storage/IFileStorage.cs

```diff
@@ -18,7 +19,61 @@ public interface IFileStorage : IHaveSerializer, IDisposable {
         Task<bool> CopyFileAsync(string path, string targetPath, CancellationToken cancellationToken = default);
         Task<bool> DeleteFileAsync(string path, CancellationToken cancellationToken = default);
         Task DeleteFilesAsync(string searchPattern = null, CancellationToken cancellation = default);
-        Task<IEnumerable<FileSpec>> GetFileListAsync(string searchPattern = null, int? limit = null, int? skip = null, CancellationToken cancellationToken = default);
+        Task<FileListResult> GetFileListAsync(string searchPattern = null, int? limit = null, CancellationToken cancellationToken = default);
+    }
+
+    public interface IHasNextPageFunc {
+        Func<Task<NextPageResult>> NextPageFunc { get; set; }
+    }
+
+    public class NextPageResult {
+        public bool Success { get; set; }
+        public bool HasMore { get; set; }
+        public IReadOnlyCollection<FileSpec> Files { get; set; }
+        public Func<Task<NextPageResult>> NextPageFunc { get; set; }
+    }
+
+    public class FileListResult : IHasNextPageFunc {
+        private static IReadOnlyCollection<FileSpec> _empty = new ReadOnlyCollection<FileSpec>(new FileSpec[0]);
+        public static FileListResult Empty = new FileListResult(_empty);
+
+        public FileListResult(IReadOnlyCollection<FileSpec> files) {
+            Files = files;
+            HasMore = false;
+            ((IHasNextPageFunc)this).NextPageFunc = null;
+        }
+        
+        public FileListResult(IReadOnlyCollection<FileSpec> files, bool hasMore, Func<Task<NextPageResult>> nextPageFunc) {
+            Files = files;
+            HasMore = hasMore;
+            ((IHasNextPageFunc)this).NextPageFunc = nextPageFunc;
+        }
+
+        public FileListResult(Func<Task<NextPageResult>> nextPageFunc) {
+            ((IHasNextPageFunc)this).NextPageFunc = nextPageFunc;
```

Comment:
I'm curious what the use cases of this would be? Also how would the current files be populated or has more?

Reply:
It’s literally being used in the code you just commented on. You give it a callback and then call nextpage to load the initial page of data. Otherwise you have to repeat yourself.